### PR TITLE
Populate internal DNS names in GCE provider

### PR DESCRIPTION
**Pull request by awly**
Both VM name and hostname are internally routable.

**What this PR does / why we need it**: GCE cloud provider only populated IP addresses for instances. This PR adds internal DNS names. DNS names are used e.g. in kubelet server certificates (https://github.com/kubernetes/kubernetes/blob/5b77996433e8e8e07a7b0a9f7436fd44da2cad9c/pkg/kubelet/kubelet.go#L385)

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by saad-ali** [comment]
LGTM

But, I'd like networking team to look at it.
/assign @bowei

**Comment by awly** [comment]
cc @mikedanese 
This was the reason we don't get DNS names in kubelet server CSRs.

**Comment by mikedanese** [comment]
Do these DNS names resolve across zones?

**Comment by awly** [comment]
Yes, these names resolve across zones within the same project

**Comment by awly** [comment]
Ping @mikedanese @bowei

**Comment by awly** [comment]
Ping @bowei

File: pkg/cloudprovider/providers/gce/gce_instances.go [hunk 1]
```diff
@@ -90,9 +90,19 @@ func (gce *GCECloud) NodeAddresses(_ context.Context, _ types.NodeName) ([]v1.No
 	if err != nil {
 		return nil, fmt.Errorf("couldn't get external IP: %v", err)
 	}
+	internalDNSFull, err := metadata.Get("instance/hostname")
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_instances.go]: is it safer to log and ignore errors getting the DNS names so the IPs can still be populated?

Comment (awly) [reply pkg/cloudprovider/providers/gce/gce_instances.go]: Done. I'm not entirely sure about the guarantees of these metadata fields being present.

**Comment by awly** [comment]
PTAL @bowei

File: pkg/cloudprovider/providers/gce/gce_instances.go [hunk 2]
```diff
@@ -91,10 +91,22 @@ func (gce *GCECloud) NodeAddresses(_ context.Context, _ types.NodeName) ([]v1.No
...
+	addresses := []v1.NodeAddress{
 		{Type: v1.NodeInternalIP, Address: internalIP},
 		{Type: v1.NodeExternalIP, Address: externalIP},
-	}, nil
+	}
+
+	if internalDNSFull, err := metadata.Get("instance/hostname"); err != nil {
+		glog.Warningf("couldn't get full internal DNS name: %v", err)
+	} else {
+		addresses = append(addresses, v1.NodeAddress{Type: v1.NodeInternalDNS, Address: internalDNSFull})
```
Comment (liggitt) [inline pkg/cloudprovider/providers/gce/gce_instances.go]: can we also add this as a Hostname address?

File: pkg/cloudprovider/providers/gce/gce_instances.go [hunk 3]
```diff
@@ -91,10 +91,22 @@ func (gce *GCECloud) NodeAddresses(_ context.Context, _ types.NodeName) ([]v1.No
...
+	}
+	if internalDNSShort, err := metadata.Get("instance/name"); err != nil {
```
Comment (mikedanese) [inline pkg/cloudprovider/providers/gce/gce_instances.go]: Let's skip shortnames until we need them.

File: pkg/cloudprovider/providers/gce/gce_instances.go [hunk 4]
```diff
@@ -91,10 +91,22 @@ func (gce *GCECloud) NodeAddresses(_ context.Context, _ types.NodeName) ([]v1.No
...
 		return nil, fmt.Errorf("couldn't get external IP: %v", err)
 	}
-	return []v1.NodeAddress{
+	addresses := []v1.NodeAddress{
 		{Type: v1.NodeInternalIP, Address: internalIP},
 		{Type: v1.NodeExternalIP, Address: externalIP},
-	}, nil
+	}
+
+	if internalDNSFull, err := metadata.Get("instance/hostname"); err != nil {
```
Comment (mikedanese) [inline pkg/cloudprovider/providers/gce/gce_instances.go]: This is the FQDN right? What does it look like?

**Review by mikedanese — COMMENTED** [review]
One comment, one question, otherwise lgtm.

/approve

**Review by awly — COMMENTED** [review]
PTAL

**Pull request merged**